api/src/main/java/edu/cornell/mannlib/vitro/webapp/dao/jena/RDFServiceGraph.java

```diff
@@ -94,37 +95,105 @@ private StringBuilder serialize(StringBuilder sb, Triple t) {
         try {
             if(!removalsGraph.isEmpty()) {
-                String removals = serializeGraph(removalsGraph);
-                changeSet.addRemoval(RDFServiceUtils.toInputStream(removals),
-                        RDFService.ModelSerializationFormat.N3, graphURI);
-                removalsGraph.clear();
+                flushGraph(removalsGraph,graphURI,changeSet,rdfService,false);
             }
             if(!additionsGraph.isEmpty()) {
-                String additions = serializeGraph(additionsGraph);
-                changeSet.addAddition(RDFServiceUtils.toInputStream(additions),
-                        RDFService.ModelSerializationFormat.N3, graphURI);
-                additionsGraph.clear();
+                flushGraph(additionsGraph,graphURI,changeSet,rdfService,true);
             }
-            rdfService.changeSetUpdate(changeSet);
         } catch (RDFServiceException rdfse) {
             throw new RuntimeException(rdfse);
         }
     }
+    
+    
+    private static int maxStringSize = 10000000;
+    
+    private synchronized void flushGraph(Graph graph,String graphURI,ChangeSet changeSet,RDFService rdfService,boolean add) throws RDFServiceException {
+        
+        Runtime runtime = Runtime.getRuntime();
 
-    private synchronized String serializeGraph(Graph graph) {
-        StringBuilder sb = new StringBuilder();
+        StringBuilder sb = new StringBuilder(maxStringSize+(10000000/10));
         Iterator<Triple> tripIt = graph.find(null, null, null);
+        
         while(tripIt.hasNext()) {
-            serialize(sb.append(" \n"), tripIt.next());
-        }
-        return sb.toString();
-    }
 
+            sb.append(" \n").append(serialize(tripIt.next()));
+           
+            if(sb.length()>maxStringSize) {
+               
+               String triples = sb.toString();
+               
+              InputStream tripleInputStream = RDFServiceUtils.toInputStream(triples);
+               
+              if(add) {
+                   changeSet.addAddition(tripleInputStream,RDFService.ModelSerializationFormat.N3, graphURI);
+              }else {
+                  changeSet.addRemoval(tripleInputStream,RDFService.ModelSerializationFormat.N3, graphURI);
+              }
+              
+               triples = null;
+               
+               rdfService.changeSetUpdate(changeSet);
+                
+               changeSet = rdfService.manufactureChangeSet();
+
+               if(tripleInputStream!=null) {
+                    try {
+                        tripleInputStream.close();
+                    } catch (IOException e) {
+                        throw new RuntimeException();
+                    }
+               }
+               
+               runtime.gc();
+               
+               sb = new StringBuilder(maxStringSize+(10000000/10));
+               
+           }
+        }
+        
+        String triples = sb.toString();
```

Comment:
Can probably refactor most of lines 168 - 192 into a separate, `private` method instead of copy/paste from line:137 above.